iree/turbine/kernel/wave/decompose_scan_ops.py

```diff
@@ -210,47 +243,77 @@ def decompose_scan_ops(
+                    f"Only implemented reduction on fastest dimension. Got {scan_dim} and {src_fastest_dims}."
+                    f"\n{custom}"
+                )
+
             get_thread_shape = lambda index: max(
                 subs_idxc(x.size) for x in index.values()
             )
 
-            try:
-                op = get_custom(scan_src)
-                thread_shape = get_thread_shape(op.index)
-                local_scan_sizes = thread_shape
-            except Exception as e:
-                index_str = "\n".join(f"{k}: {v}" for k, v in op.index.items())
-                raise RuntimeError(
-                    f"Error in decompose_scan_ops: {scan_src} with index\n"
-                    f"{index_str}\n{scan_src=}\n{scan_acc=}\n{scan_dim=}"
-                ) from e
-
-            local_scan = [scan_src]
-
-            if local_scan_sizes > 1:
-                local_scan = emit_local_inclusive_scan(
-                    binary_fn, scan_src, custom.graph, local_scan_sizes
+            local_scan_sizes = []
+            for arg in scan_src:
+                try:
+                    op = get_custom(arg)
+                    thread_shape = get_thread_shape(op.index)
+                    local_scan_sizes.append(thread_shape)
+                except Exception as e:
+                    index_str = "\n".join(f"{k}: {v}" for k, v in op.index.items())
+                    raise RuntimeError(
+                        f"Error in decompose_scan_ops: {scan_src} with index\n"
+                        f"{index_str}\n{scan_src=}\n{scan_acc=}\n{scan_dim=}"
+                    ) from e
+
+            if not all_equal(local_scan_sizes):
+                raise NotImplementedError(
+                    "NYI: Expect all reduce_src to have same local reduce size."
                 )
 
-            result = emit_global_scan(
+            # Phase 1: Local scan
+            local_scan = emit_local_inclusive_scan(
+                binary_fn, scan_src, custom.graph, local_scan_sizes[0]
+            )
+
+            # Phase 2: Src wide local scan
+            local_scan = emit_variable_scan(
+                binary_fn, local_scan, custom.graph, local_scan_sizes[0]
+            )
+
+            global_scan = emit_global_scan(
                 binary_fn,
-                scan_src,
+                scan_src[0],
                 local_scan,
                 custom.graph,
                 subgroup_size,
                 hardware_constraint,
-                local_scan_sizes,
+                local_scan_sizes[0],
                 scan_dim,
             )
 
-            custom.replace_all_uses_with(result)
+            # Update the users based on the global scan `reshape` results.
+            users = [
+                (user, user.fx_node.args.index(custom.fx_node)) for user in custom.users
+            ]
+
+            for user, arg_index in users:
+                user.update_arg(arg_index, global_scan[user.expanded_dims[scan_dim]])
```

Comment:
Can you update `CustomOp.update_arg` to accept `fx.Node` instead of calculating index here?